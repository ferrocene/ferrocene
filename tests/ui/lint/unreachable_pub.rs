//@ check-pass

#![allow(unused)]
#![warn(unreachable_pub)]

mod private_mod {
    // non-leaked `pub` items in private module should be linted
    pub use std::fmt; //~ WARNING unreachable_pub
    pub use std::env::{Args}; // braced-use has different item spans than unbraced
    //~^ WARNING unreachable_pub

    // we lint on struct definition
    pub struct Hydrogen { //~ WARNING unreachable_pub
        // but not on fields, even if they are `pub` as putting `pub(crate)`
        // it would clutter the source code for little value
        pub neutrons: usize,
        pub(crate) electrons: usize
    }
    pub(crate) struct Calcium {
        pub neutrons: usize,
    }
    impl Hydrogen {
        // impls, too
        pub fn count_neutrons(&self) -> usize { self.neutrons } //~ WARNING unreachable_pub
        pub(crate) fn count_electrons(&self) -> usize { self.electrons }
    }
    impl Clone for Hydrogen {
        fn clone(&self) -> Hydrogen {
            Hydrogen { neutrons: self.neutrons, electrons: self.electrons }
        }
    }

    pub enum Helium {} //~ WARNING unreachable_pub
    pub union Lithium { c1: usize, c2: u8 } //~ WARNING unreachable_pub
    pub fn beryllium() {} //~ WARNING unreachable_pub
    pub trait Boron {} //~ WARNING unreachable_pub
    pub const CARBON: usize = 1; //~ WARNING unreachable_pub
    pub static NITROGEN: usize = 2; //~ WARNING unreachable_pub
    pub type Oxygen = bool; //~ WARNING unreachable_pub

    macro_rules! define_empty_struct_with_visibility {
        ($visibility: vis, $name: ident) => { $visibility struct $name {} }
        //~^ WARNING unreachable_pub
    }
    define_empty_struct_with_visibility!(pub, Fluorine);

    extern "C" {
        pub fn catalyze() -> bool; //~ WARNING unreachable_pub
    }

    // items leaked through signatures (see `get_neon` below) are OK
    pub struct Neon {}

    // crate-visible items are OK
    pub(crate) struct Sodium {}
}

pub mod public_mod {
    // module is public: these are OK, too
    pub struct Magnesium {}
    pub(crate) struct Aluminum {}
}

pub fn get_neon() -> private_mod::Neon {
    private_mod::Neon {}
}

fn main() {
    let _ = get_neon();
}
