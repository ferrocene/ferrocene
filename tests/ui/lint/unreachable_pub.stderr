warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:8:13
   |
LL |     pub use std::fmt;
   |     ---     ^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates
note: the lint level is defined here
  --> $DIR/unreachable_pub.rs:4:9
   |
LL | #![warn(unreachable_pub)]
   |         ^^^^^^^^^^^^^^^

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:9:24
   |
LL |     pub use std::env::{Args}; // braced-use has different item spans than unbraced
   |     ---                ^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:13:5
   |
LL |     pub struct Hydrogen {
   |     ---^^^^^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:24:9
   |
LL |         pub fn count_neutrons(&self) -> usize { self.neutrons }
   |         ---^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         help: consider restricting its visibility: `pub(crate)`

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:33:5
   |
LL |     pub enum Helium {}
   |     ---^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:34:5
   |
LL |     pub union Lithium { c1: usize, c2: u8 }
   |     ---^^^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:35:5
   |
LL |     pub fn beryllium() {}
   |     ---^^^^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:36:5
   |
LL |     pub trait Boron {}
   |     ---^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:37:5
   |
LL |     pub const CARBON: usize = 1;
   |     ---^^^^^^^^^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:38:5
   |
LL |     pub static NITROGEN: usize = 2;
   |     ---^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:39:5
   |
LL |     pub type Oxygen = bool;
   |     ---^^^^^^^^^^^^
   |     |
   |     help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:42:47
   |
LL |         ($visibility: vis, $name: ident) => { $visibility struct $name {} }
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
LL |     define_empty_struct_with_visibility!(pub, Fluorine);
   |     ---------------------------------------------------
   |     |                                    |
   |     |                                    help: consider restricting its visibility: `pub(crate)`
   |     in this macro invocation
   |
   = help: or consider exporting it for use by other crates
   = note: this warning originates in the macro `define_empty_struct_with_visibility` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unreachable `pub` item
  --> $DIR/unreachable_pub.rs:48:9
   |
LL |         pub fn catalyze() -> bool;
   |         ---^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         help: consider restricting its visibility: `pub(crate)`
   |
   = help: or consider exporting it for use by other crates

warning: 13 warnings emitted

